### PR TITLE
nixos/tests: add check mariadb galera cluster

### PR DESCRIPTION
**Pull request by Izorkin**
###### Motivation for this change
Moved mysql tests to subfolder.
Added check mariadb galera cluster.

cc @flokli @aanderse @worldofpeace @disassembler.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Izorkin** [comment]
@GrahamcOfBorg test mariadbGalera

**Comment by Izorkin** [comment]
@GrahamcOfBorg test mariadbGalera

**Comment by Izorkin** [comment]
@GrahamcOfBorg test mariadbGalera

File: nixos/tests/all-tests.nix [hunk 1]
```diff
@@ -172,6 +172,7 @@ in
   magnetico = handleTest ./magnetico.nix {};
   magic-wormhole-mailbox-server = handleTest ./magic-wormhole-mailbox-server.nix {};
   mailcatcher = handleTest ./mailcatcher.nix {};
+  mariadbGalera = handleTest ./mysql/mariadb-galera.nix {};
```
Comment (flokli) [inline nixos/tests/all-tests.nix:175]:
```suggestion
  mariadb-galera = handleTest ./mysql/mariadb-galera.nix {};
```

File: nixos/tests/all-tests.nix [hunk 2]
```diff
@@ -193,9 +194,9 @@ in
   munin = handleTest ./munin.nix {};
   mutableUsers = handleTest ./mutable-users.nix {};
   mxisd = handleTest ./mxisd.nix {};
-  mysql = handleTest ./mysql.nix {};
-  mysqlBackup = handleTest ./mysql-backup.nix {};
-  mysqlReplication = handleTest ./mysql-replication.nix {};
+  mysql = handleTest ./mysql/mysql.nix {};
+  mysqlBackup = handleTest ./mysql/mysql-backup.nix {};
```
Comment (flokli) [inline nixos/tests/all-tests.nix:198]:
```suggestion
  mysql-backup = handleTest ./mysql/mysql-backup.nix {};
```

File: nixos/tests/all-tests.nix [hunk 3]
```diff
@@ -193,9 +194,9 @@ in
...
+  mysqlReplication = handleTest ./mysql/mysql-replication.nix {};
```
Comment (flokli) [inline nixos/tests/all-tests.nix:199]:
```suggestion
  mysql-replication = handleTest ./mysql/mysql-replication.nix {};
```

File: nixos/tests/mysql/mariadb-galera.nix [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+  mysqlenv-mariabackup = pkgs.buildEnv { name = "mysql-path-env-mariabackup"; pathsToLink = [ "/bin" ]; paths = with pkgs; [ gzip iproute netcat procps pv socat ]; };
+  mysqlenv-rsync       = pkgs.buildEnv { name = "mysql-path-env-rsync";       pathsToLink = [ "/bin" ]; paths = with pkgs; [ lsof procps rsync stunnel ]; };
+
+in {
+  name = "mariadb-galera";
+  meta = with pkgs.stdenv.lib.maintainers; {
+    maintainers = [ izorkin ];
+  };
+
+  nodes = {
```
Comment (flokli) [inline nixos/tests/mysql/mariadb-galera.nix:14]: I'm not really that familiar with Galera. Anyone around that can take a look at this?

File: nixos/tests/mysql/mariadb-galera.nix [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+        ensureDatabases = [ "testdb" ];
+        ensureUsers = [{
+          name = "testuser";
+          ensurePermissions = {
+            "testdb.*" = "ALL PRIVILEGES";
+          };
+        }];
+        initialScript = pkgs.writeText "mariadb-init.sql" ''
+          GRANT ALL PRIVILEGES ON *.* TO 'check_repl'@'localhost' IDENTIFIED BY 'check_pass' WITH GRANT OPTION;
+          GRANT ALL PRIVILEGES ON *.* TO 'check_repl'@'192.168.1.1' IDENTIFIED BY 'check_pass' WITH GRANT OPTION;
```
Comment (flokli) [inline nixos/tests/mysql/mariadb-galera.nix:44]: isn't that equivalent to localhost? if it's required due to other reasons, probably worth adding a comment.

File: nixos/tests/mysql/mariadb-galera.nix [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+          GRANT ALL PRIVILEGES ON *.* TO 'check_repl'@'192.168.1.2' IDENTIFIED BY 'check_pass' WITH GRANT OPTION;
```
Comment (flokli) [inline nixos/tests/mysql/mariadb-galera.nix:45]: Other tests seem to refer to these IPs via things like
```
(lib.head nodes.galera_02_maria.config.networking.interfaces.eth1.ipv4.addresses)
```

I'm not sure if it gets too messy - maybe we can add these lines to `/etc/hosts` and use hostnames here?
> (lib.head config.networking.interfaces.eth1.ipv4.addresses).address

File: nixos/tests/mysql/mariadb-galera.nix (hunk 4 above)
Comment (flokli) [reply nixos/tests/mysql/mariadb-galera.nix:14]: Also, we might want to add a textual comment describing how the cluster is set up and what things are tested, so others can see what this test tries to test just by looking at the comment.

File: nixos/tests/mysql/mariadb-galera.nix (hunk 5 above)
Comment (Izorkin) [reply nixos/tests/mysql/mariadb-galera.nix:44]: Rechecked, dont` need.

File: nixos/tests/mysql/mariadb-galera.nix (hunk 4 above)
Comment (Izorkin) [reply nixos/tests/mysql/mariadb-galera.nix:18]: Added comment.

**Comment by Izorkin** [comment]
@GrahamcOfBorg test mariadbGalera

**Comment by Izorkin** [comment]
@GrahamcOfBorg test mariadb-galera

**Comment by Izorkin** [comment]
@GrahamcOfBorg test mariadb-galera

**Comment by Izorkin** [comment]
@GrahamcOfBorg eval

**Comment by flokli** [comment]
@Izorkin Ran the test locally, it seems it's stuck at

```
galera_03_rsync # [  201.426407] mysqld[893]: 2020-04-04 22:46:55 0 [Note] WSREP: resuming provider at 10
galera_03_rsync # [  201.429545] mysqld[893]: 2020-04-04 22:46:55 0 [Note] WSREP: Provider resumed.
galera_04_rsync # [  142.918417] rsyncd[1436]: name lookup failed for 192.168.2.1: Temporary failure in name resolution
galera_04_rsync # [  142.920801] rsyncd[1436]: connect from UNKNOWN (192.168.2.1)
galera_04_rsync # [  142.924294] rsyncd[1436]: rsync to rsync_sst/ from UNKNOWN (192.168.2.1)
galera_04_rsync # [  142.925897] rsyncd[1436]: receiving file list
galera_04_rsync # [  142.972423] rsyncd[1436]: sent 48 bytes  received 193 bytes  total size 41
galera_03_rsync # [  216.515459] mysqld[893]: 2020-04-04 22:47:10 0 [Note] WSREP: 0.0 (galera_03): State transfer to 1.0 (galera_04) complete.
galera_04_rsync # [  142.978348] mysqld[903]: 2020-04-04 22:47:10 0 [Note] WSREP: 0.0 (galera_03): State transfer to 1.0 (galera_04) complete.
galera_03_rsync # [  216.517765] mysqld[893]: 2020-04-04 22:47:10 0 [Note] WSREP: Shifting DONOR/DESYNCED -> JOINED (TO: 5)
galera_03_rsync # [  216.521805] mysqld[893]: 2020-04-04 22:47:10 0 [Note] WSREP: Member 0.0 (galera_03) synced with group.
galera_03_rsync # [  216.523508] mysqld[893]: 2020-04-04 22:47:10 0 [Note] WSREP: Shifting JOINED -> SYNCED (TO: 5)
galera_04_rsync # [  142.983994] mysqld[903]: 2020-04-04 22:47:10 0 [Note] WSREP: Member 0.0 (galera_03) synced with group.
galera_03_rsync # [  216.525931] mysqld[893]: 2020-04-04 22:47:10 1 [Note] WSREP: Synchronized with group, ready for connections
galera_03_rsync # [  216.527844] mysqld[893]: 2020-04-04 22:47:10 1 [Note] WSREP: wsrep_notify_cmd is not defined, skipping notification.
galera_04_rsync # [  143.125731] mysqld[903]: WSREP_SST: [INFO] Joiner cleanup. rsync PID: 961 (20200404 22:47:10.567)
galera_04_rsync # [  143.528222] rsyncd[961]: sent 0 bytes  received 0 bytes  total size 0
galera_04_rsync # [  143.670738] mysqld[903]: WSREP_SST: [INFO] Joiner cleanup done. (20200404 22:47:11.112)
galera_04_rsync # [  143.678394] mysqld[903]: 2020-04-04 22:47:11 0 [Note] WSREP: SST complete, seqno: 5
galera_04_rsync # [  143.703722] mysqld[903]: 2020-04-04 22:47:11 0 [Note] InnoDB: Using Linux native AIO
galera_04_rsync # [  143.705639] mysqld[903]: 2020-04-04 22:47:11 0 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins
galera_04_rsync # [  143.707979] mysqld[903]: 2020-04-04 22:47:11 0 [Note] InnoDB: Uses event mutexes
galera_04_rsync # [  143.709711] mysqld[903]: 2020-04-04 22:47:11 0 [Note] InnoDB: Compressed tables use zlib 1.2.11
galera_04_rsync # [  143.711215] mysqld[903]: 2020-04-04 22:47:11 0 [Note] InnoDB: Number of pools: 1
galera_04_rsync # [  143.712594] mysqld[903]: 2020-04-04 22:47:11 0 [Note] InnoDB: Using SSE2 crc32 instructions
galera_04_rsync # [  143.718026] mysqld[903]: 2020-04-04 22:47:11 0 [Note] InnoDB: Initializing buffer pool, total size = 128M, instances = 1, chunk size = 128M
galera_04_rsync # [  143.721941] mysqld[903]: 2020-04-04 22:47:11 0 [Note] InnoDB: Completed initialization of buffer pool
galera_04_rsync # [  143.724946] mysqld[903]: 2020-04-04 22:47:11 0 [Note] InnoDB: If the mysqld execution user is authorized, page cleaner thread priority can be changed. See the man page of setpriority().
galera_04_rsync # [  143.730072] mysqld[903]: 2020-04-04 22:47:11 0 [Note] InnoDB: Starting crash recovery from checkpoint LSN=1635075
galera_04_rsync # [  143.838651] mysqld[903]: 2020-04-04 22:47:11 0 [Note] InnoDB: 128 out of 128 rollback segments are active.
galera_04_rsync # [  143.841085] mysqld[903]: 2020-04-04 22:47:11 0 [Note] InnoDB: Creating shared tablespace for temporary tables
galera_04_rsync # [  143.844787] mysqld[903]: 2020-04-04 22:47:11 0 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
galera_04_rsync # [  143.849868] mysqld[903]: 2020-04-04 22:47:11 0 [Note] InnoDB: File './ibtmp1' size is now 12 MB.
galera_04_rsync # [  143.852109] mysqld[903]: 2020-04-04 22:47:11 0 [Note] InnoDB: Waiting for purge to start
galera_04_rsync # [  143.904178] mysqld[903]: 2020-04-04 22:47:11 0 [Note] InnoDB: 10.3.22 started; log sequence number 1635084; transaction id 33
galera_04_rsync # [  143.907945] mysqld[903]: 2020-04-04 22:47:11 0 [Note] Plugin 'FEEDBACK' is disabled.
galera_03_rsync # [  217.454788] mysqld[893]: 2020-04-04 22:47:11 0 [Note] WSREP: 1.0 (galera_04): State transfer from 0.0 (galera_03) complete.
galera_04_rsync # [  143.914193] mysqld[903]: 2020-04-04 22:47:11 0 [Note] InnoDB: Loading buffer pool(s) from /var/lib/mysql/ib_buffer_pool
galera_03_rsync # [  217.456750] mysqld[893]: 2020-04-04 22:47:11 0 [Note] WSREP: Member 1.0 (galera_04) synced with group.
galera_04_rsync # [  143.919824] mysqld[903]: 2020-04-04 22:47:11 0 [Note] InnoDB: Buffer pool(s) load completed at 200404 22:47:11
galera_04_rsync # [  143.922910] mysqld[903]: 2020-04-04 22:47:11 0 [Note] Server socket created on IP: '0.0.0.0'.
galera_04_rsync # [  143.925181] mysqld[903]: 2020-04-04 22:47:11 0 [Note] WSREP: Signalling provider to continue.
galera_04_rsync # [  143.927311] mysqld[903]: 2020-04-04 22:47:11 0 [Note] WSREP: SST received: c2866648-76c5-11ea-af3c-e751edcdf91b:5
galera_04_rsync # [  143.929547] mysqld[903]: 2020-04-04 22:47:11 0 [Note] WSREP: 1.0 (galera_04): State transfer from 0.0 (galera_03) complete.
galera_04_rsync # [  143.931665] mysqld[903]: 2020-04-04 22:47:11 0 [Note] WSREP: Shifting JOINER -> JOINED (TO: 5)
galera_04_rsync # [  143.933587] mysqld[903]: 2020-04-04 22:47:11 0 [Note] WSREP: Member 1.0 (galera_04) synced with group.
galera_04_rsync # [  143.935348] mysqld[903]: 2020-04-04 22:47:11 0 [Note] WSREP: Shifting JOINED -> SYNCED (TO: 5)
galera_04_rsync # [  143.937376] mysqld[903]: 2020-04-04 22:47:11 1 [Note] WSREP: Synchronized with group, ready for connections
galera_04_rsync # [  143.939631] mysqld[903]: 2020-04-04 22:47:11 1 [Note] WSREP: wsrep_notify_cmd is not defined, skipping notification.
galera_04_rsync # [  143.942303] mysqld[903]: 2020-04-04 22:47:11 0 [Note] Reading of all Master_info entries succeeded
galera_04_rsync # [  143.944319] mysqld[903]: 2020-04-04 22:47:11 0 [Note] Added new Master_info '' to hash table
galera_04_rsync # [  143.946434] mysqld[903]: 2020-04-04 22:47:11 0 [Note] /nix/store/kfiajz1r1cwxf5j9kmx2ma5cm42bf1l1-mariadb-server-10.3.22/bin/mysqld: ready for connections.
galera_04_rsync # [  143.949073] mysqld[903]: Version: '10.3.22-MariaDB'  socket: '/run/mysqld/mysqld.sock'  port: 3306  MariaDB Server
galera_04_rsync # [  143.988935] systemd[1]: Started MySQL Server.
galera_04_rsync # [  143.991279] systemd[1]: Reached target Multi-User System.
galera_04_rsync # [  143.992629] systemd[1]: Startup finished in 5.027s (kernel) + 2min 18.960s (userspace) = 2min 23.988s.
```

did you run it via `nixosTests.mariadb-galera`?

Also, I noticed a lot of spam about temporary failure in name resolution, while trying to do a reverse lookup of the internal IPs.

Could this be improved it we add the IPs and names to `/etc/hosts`, like it's done in ` nixos/tests/nfs/kerberos.nix`?

Also, doing that, we might be able to switch some of the IPs to hostnames.

**Comment by Izorkin** [comment]
> did you run it via `nixosTests.mariadb-galera`?
> 
Yes, it works for me

Added hostnames.

**Comment by Izorkin** [comment]
@GrahamcOfBorg test mariadb-galera

File: nixos/tests/all-tests.nix [hunk 7]
```diff
@@ -193,9 +193,10 @@ in
   munin = handleTest ./munin.nix {};
   mutableUsers = handleTest ./mutable-users.nix {};
   mxisd = handleTest ./mxisd.nix {};
-  mysql = handleTest ./mysql.nix {};
-  mysqlBackup = handleTest ./mysql-backup.nix {};
-  mysqlReplication = handleTest ./mysql-replication.nix {};
+  mysql = handleTest ./mysql/mysql.nix {};
+  mysql-autobackup = handleTest ./mysql/mysql-autobackup.nix {};
```
Comment (flokli) [inline nixos/tests/all-tests.nix:197]: Can you put the rename of the test attributes into the first commit (the one that also moves files around), and the addition of the galera test into a second one?

File: nixos/tests/mysql/mariadb-galera.nix [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+      { pkgs, ... }:
+      {
+      networking = {
+        interfaces.eth1 = {
+          ipv4.addresses = [
+            { address = "192.168.1.1"; prefixLength = 24; }
+          ];
+        };
+        extraHosts = ''
+          192.168.1.1 galera_01
```
Comment (flokli) [inline nixos/tests/mysql/mariadb-galera.nix:29]: If galera_01/02 and galera03/04 never actually talk to each other, maybe make it separate VM tests?

**Comment by Izorkin** [comment]
@GrahamcOfBorg test mariadb-galera-mariabackup
@GrahamcOfBorg test mariadb-galera-rsync

**Comment by flokli** [comment]
Pinging some more people controbuting to mysql for some more feedback.

File: pkgs/servers/sql/mariadb/default.nix [hunk 9]
```diff
@@ -101,6 +101,15 @@ common = rec { # attributes common to both builds
 
   passthru.mysqlVersion = "5.7";
 
+  passthru.tests = {
+    mariadb-galera-mariabackup = nixosTests.mariadb-galera-mariabackup;
```
Comment (Mic92) [inline pkgs/servers/sql/mariadb/default.nix:105]: YOu also need to add `nixosTests` as an argument.

Comment (Izorkin) [reply pkgs/servers/sql/mariadb/default.nix:105]: Fixed.

File: nixos/tests/all-tests.nix [hunk 10]
```diff
@@ -197,9 +198,10 @@ in
...
```
Comment (Mic92) [inline nixos/tests/all-tests.nix:202]: This file does unfortunately not exists. You renamed it to `./mysql/automysqlbackup.nix`.

Comment (Izorkin) [reply nixos/tests/all-tests.nix:202]: Oops, fixed.

**Comment by Izorkin** [comment]
@GrahamcOfBorg test mariadb-galera-mariabackup
@GrahamcOfBorg test mariadb-galera-rsync
@GrahamcOfBorg test mysql
@GrahamcOfBorg test mysql-autobackup
@GrahamcOfBorg test mysql-backup
@GrahamcOfBorg test mysql-replication

**Pull request merged**